File: balance_controller.py
"""
VMC + LQR 轮腿平衡控制器 (精简版)
"""
import mujoco as mj, numpy as np, time, math, mujoco.viewer
from robot_config import (
    init_model, fk_deriv, ik, vmc_jacobian,
    phi_to_q, q_to_phi, phi_dot,
    get_pitch, get_pitch_dot, get_acc, get_yaw, get_yaw_dot,
    lookup_k_mat, build_k_table,
    M, G, MW,
)

model, data, act_ids, act_idx, joint_addr, freejoint_adr, wheel_body = \
    init_model("COD-2026RoboMaster-Balance copy.xml")

L_REF = 0.30; KP_L=1500.0; KD_L=400.0; TLIM=100.0
# 转向环 (yaw) PD 增益
KP_YAW = 30.0; KD_YAW = 1.0
WP = (1, -1)  # wheel_pol R,L
VS = ((1,-1),(-1,1))  # VMC_SIGN R,L
FIXED_BODY = freejoint_adr < 0  # 机体是否被固定

class LD:
    def __init__(s): s.FN=100.0; s.g=True; s.th=s.td=0.0; s.Tr=s.Tf=s.Tw=0.0

Rl, Ll = LD(), LD()

def main():
    p1r, p4r = ik(L_REF)
    qRr,qRf = phi_to_q("Right",p1r,p4r); qLr,qLf = phi_to_q("Left",p1r,p4r)
    data.qpos[joint_addr["Right"]["rear"]["qpos"]]  = qRr
    data.qpos[joint_addr["Right"]["front"]["qpos"]] = qRf
    data.qpos[joint_addr["Left"]["rear"]["qpos"]]   = qLr
    data.qpos[joint_addr["Left"]["front"]["qpos"]]  = qLf
    mj.mj_forward(model, data)
    build_k_table()
    x0_body = data.qpos[freejoint_adr] if not FIXED_BODY else 0.0  # 初始 x (里程计零点)
    yaw0 = get_yaw(data, freejoint_adr) if not FIXED_BODY else 0.0   # 目标 yaw
    lp = time.time()

    with mujoco.viewer.launch_passive(model, data) as v:
        while v.is_running():
            if FIXED_BODY:
                pitch = 0.0; pitch_d = 0.0; acc = [0.0, 0.0, 0.0]
            else:
                pitch = get_pitch(data, freejoint_adr)
                pitch_d = get_pitch_dot(data, freejoint_adr)
                acc = get_acc(data, freejoint_adr)

            # 里程计: 用 freejoint 绝对位置（仿真真值），避免轮子编码器积分误差
            if not FIXED_BODY:
                body_x  = data.qpos[freejoint_adr] - x0_body
                body_xd = data.qvel[freejoint_adr]
                yaw     = get_yaw(data, freejoint_adr)
                yaw_d   = get_yaw_dot(data, freejoint_adr)
            else:
                body_x = 0.0; body_xd = 0.0
                yaw = 0.0; yaw_d = 0.0

            # 转向环: yaw 偏离时左右轮差速
            yaw_err = yaw - yaw0
            Tyaw = KP_YAW * yaw_err + KD_YAW * yaw_d

            for side, ld, wp, (rs,fs) in [("Right",Rl,WP[0],VS[0]),("Left",Ll,WP[1],VS[1])]:
                a = joint_addr[side]
                qr,qdr = data.qpos[a["rear"]["qpos"]],  data.qvel[a["rear"]["dof"]]
                qf,qdf = data.qpos[a["front"]["qpos"]], data.qvel[a["front"]["dof"]]
                p1,p4 = q_to_phi(side,qr,qf); p1d,p4d = phi_dot(side,qdr,qdf)
                L,th,thd,Ld,p2,p3,p5,_ = fk_deriv(p1,p4,p1d,p4d,pitch,pitch_d)
                H=L*math.cos(th); Hd=Ld*math.cos(th)-L*math.sin(th)*thd
                #腿长控制PID + 重力前馈
                FL_FF = 50.0  # 前馈力 [N]
                FL = 0.5*M*G*math.cos(th) + FL_FF + KP_L*(L_REF-H) + KD_L*(0-Hd)
                yMdd=acc[0]*math.sin(pitch)+acc[2]*math.cos(pitch)
                ld.FN=MW*yMdd+FL*math.cos(th)+MW*G; ld.g=ld.FN>=20.0
                K=lookup_k_mat(L,ld.g)
                # 反馈矩阵
                u=-K@np.array([th,thd,body_x,body_xd,-pitch,-pitch_d])
                TL=u[1]
                J=vmc_jacobian(p1,p2,p3,p4,p5,L); Tj=J@[FL,TL]
                Tr=rs*(-Tj[0]); Tf=fs*(-Tj[1]); Tw=wp*u[0]; ld.Tr=Tr; ld.Tf=Tf; ld.Tw=Tw
                data.ctrl[act_ids[act_idx[side,"rear"]]]  = np.clip(Tr,-TLIM,TLIM)
                data.ctrl[act_ids[act_idx[side,"front"]]] = np.clip(Tf,-TLIM,TLIM)
                Tyaw_sign = -1 if side=="Right" else 1 #转向环
                data.ctrl[act_ids[2 if side=="Right" else 5]] = np.clip(Tw + Tyaw_sign * Tyaw, -TLIM, TLIM)
                ld.th=th; ld.td=thd

            # ---- 扰动: 定时对轮子施加水平力 ----
            # if FIXED_BODY:
            #     t = data.time
            #     cycle = t % PERT_INTERVAL
            #     if cycle < PERT_DURATION and t - last_pert_time > PERT_DURATION:
            #         last_pert_time = t
            #     if last_pert_time > 0 and t - last_pert_time < PERT_DURATION:
            #         sign = 1.0 if (int(t / PERT_INTERVAL) % 2 == 0) else -1.0
            #         fx = sign * PERT_FORCE
            #         data.xfrc_applied[wheel_body["Right"]][0] = fx
            #         data.xfrc_applied[wheel_body["Left"]][0]  = fx
            #     else:
            #         data.xfrc_applied[wheel_body["Right"]][0] = 0.0
            #         data.xfrc_applied[wheel_body["Left"]][0]  = 0.0

            mj.mj_step(model, data); v.sync(); time.sleep(0.001)
            now=time.time()
            if now-lp>=1.0:
                print(f"[t={data.time:.1f}s] pitch={math.degrees(pitch):.1f} yaw={math.degrees(yaw):.1f} yaw_tq={Tyaw:+.1f}  FN_R={Rl.FN:.0f}N FN_L={Ll.FN:.0f}N")
                print(f"  R: Tr={Rl.Tr:+.1f} Tf={Rl.Tf:+.1f} Tw={Rl.Tw:+.1f}  th={math.degrees(Rl.th):.1f}")
                print(f"  L: Tr={Ll.Tr:+.1f} Tf={Ll.Tf:+.1f} Tw={Ll.Tw:+.1f}  th={math.degrees(Ll.th):.1f}")
                lp=now

if __name__=="__main__": main()

File: c_controller/balance_controller.c
/*
 * VMC + LQR 轮腿平衡控制器 (C 版本)
 * ====================================
 * F_Leg 腿长PD + T_Leg(LQR) → VMC Jᵀ → T_rear/T_front
 * LQR → T_wheel
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "robot_config.h"

// ======================== 参数 ========================
#define XML_PATH       "COD-2026RoboMaster-Balance copy.xml"
#define LEG_LENGTH_REF 0.20
#define KP_LEG         300.0
#define KD_LEG         100.0
#define KP_THETA       6.0
#define KD_THETA       0.2
#define TORQUE_LIMIT   100.0

// ======================== 每条腿状态 ========================
typedef struct {
    double x, x0;
    double FN;
    int    grounded;
    double theta, theta_dot;
} LegData;

// ======================== MuJoCo 句柄 ========================
static mjModel *m = NULL;
static mjData  *d = NULL;
static int act_ids[6];
static int act_idx_map[4];      // R_rear,R_front,L_rear,L_front
static int jnt_qr[2], jnt_qf[2];
static int jnt_dr[2], jnt_df[2];
static int jnt_qw[2], jnt_dw[2];
static int freejoint_adr;

// ======================== 初始化 ========================
static int init_model(const char *xml_path) {
    char err[1000] = {0};
    m = mj_loadXML(xml_path, NULL, err, sizeof(err));
    if (!m) { fprintf(stderr, "ERROR: load XML '%s'\n  %s\n", xml_path, err); return -1; }
    d = mj_makeData(m);
    if (!d) { fprintf(stderr, "ERROR: mj_makeData\n"); mj_deleteModel(m); m=NULL; return -1; }

    const char *anames[6] = {
        "Right_front_joint_actuator","Right_rear_joint_actuator","Right_Wheel_joint_actuator",
        "Left_front_joint_actuator", "Left_rear_joint_actuator", "Left_Wheel_joint_actuator",
    };
    for (int i=0;i<6;i++) {
        act_ids[i] = mj_name2id(m, mjOBJ_ACTUATOR, anames[i]);
        if (act_ids[i] < 0) { fprintf(stderr, "ERROR: actuator '%s'\n", anames[i]); return -1; }
    }
    act_idx_map[0]=1; act_idx_map[1]=0; act_idx_map[2]=4; act_idx_map[3]=3;

    #define GET_J(name, q, dv) { \
        int jid=mj_name2id(m,mjOBJ_JOINT,name); \
        if (jid<0) { fprintf(stderr,"ERROR: joint '%s'\n",name); return -1; } \
        *(q)=m->jnt_qposadr[jid]; *(dv)=m->jnt_dofadr[jid]; }
    GET_J("Right_rear_joint",  &jnt_qr[0],&jnt_dr[0]);
    GET_J("Right_front_joint", &jnt_qf[0],&jnt_df[0]);
    GET_J("Right_Wheel_joint", &jnt_qw[0],&jnt_dw[0]);
    GET_J("Left_rear_joint",   &jnt_qr[1],&jnt_dr[1]);
    GET_J("Left_front_joint",  &jnt_qf[1],&jnt_df[1]);
    GET_J("Left_Wheel_joint",  &jnt_qw[1],&jnt_dw[1]);
    #undef GET_J

    int jid = mj_name2id(m, mjOBJ_JOINT, "base_freejoint");
    freejoint_adr = (jid >= 0) ? m->jnt_qposadr[jid] : -1;
    return 0;
}

// ======================== 控制回调 ========================
static LegData leg_R = {0,0,100.0,1,0,0};
static LegData leg_L = {0,0,100.0,1,0,0};

void my_controller(const mjModel *m_model, mjData *d_data) {
    (void)m_model;
    double pitch=0,pitch_d=0,acc[3]={0,0,0};
    if (freejoint_adr>=0) {
        pitch=get_pitch(d_data,freejoint_adr);
        pitch_d=get_pitch_dot(d_data,freejoint_adr);
        get_acc(d_data,freejoint_adr,acc);
    }

    double wheel_pol[2]={1.0,-1.0};
    double S[2][13];

    // ==== pass 1: FK+LQR ====
    for(int leg=0;leg<2;leg++){
        int isR=(leg==0);
        LegData *ld=isR?&leg_R:&leg_L;
        double qr=d_data->qpos[isR?jnt_qr[0]:jnt_qr[1]];
        double qdr=d_data->qvel[isR?jnt_dr[0]:jnt_dr[1]];
        double qf=d_data->qpos[isR?jnt_qf[0]:jnt_qf[1]];
        double qdf=d_data->qvel[isR?jnt_df[0]:jnt_df[1]];
        double qw=d_data->qpos[isR?jnt_qw[0]:jnt_qw[1]];
        double qdw=d_data->qvel[isR?jnt_dw[0]:jnt_dw[1]];

        double phi1,phi4,phi1_d,phi4_d;
        q_to_phi(isR,qr,qf,&phi1,&phi4);
        phi_dot(isR,qdr,qdf,&phi1_d,&phi4_d);

        double Lv,theta,theta_d,L_d,phi2,phi3,phi5;
        fk_deriv(phi1,phi4,phi1_d,phi4_d,pitch,pitch_d,0.0001,
                 &Lv,&theta,&theta_d,&L_d,&phi2,&phi3,&phi5);

        if(ld->x0==0&&qw!=0)ld->x0=qw;
        ld->x=WHEEL_R*(qw-ld->x0);
        double x_dot=WHEEL_R*qdw;

        double H=Lv*cos(theta), Hd=L_d*cos(theta)-Lv*sin(theta)*theta_d;
        double F_Leg=0.5*M_BODY*GRAVITY*cos(theta)
                    +KP_LEG*(LEG_LENGTH_REF-H)+KD_LEG*(0-Hd);

        double yMdd=acc[0]*sin(pitch)+acc[2]*cos(pitch);
        double P_est=F_Leg*cos(theta);
        ld->FN=MW*yMdd+P_est+MW*GRAVITY;
        ld->grounded=(ld->FN>=20.0);

        double K[2][6];
        lookup_k_mat(Lv,ld->grounded,K);
        double X[6]={theta,theta_d,ld->x,x_dot,pitch,pitch_d};
        double u[2]={0,0};
        for(int i=0;i<2;i++)for(int j=0;j<6;j++)u[i]+=K[i][j]*(-X[j]);

        double P=F_Leg*cos(theta)+u[1]*sin(theta)/(Lv>0.01?Lv:0.01);
        ld->FN=MW*yMdd+P+MW*GRAVITY;
        ld->theta=theta; ld->theta_dot=theta_d;

        S[leg][0]=phi1;S[leg][1]=phi4;S[leg][2]=phi1_d;S[leg][3]=phi4_d;
        S[leg][4]=Lv;S[leg][5]=theta;S[leg][6]=theta_d;
        S[leg][7]=phi2;S[leg][8]=phi3;S[leg][9]=phi5;
        S[leg][10]=u[0];S[leg][11]=u[1];S[leg][12]=F_Leg;
    }

    // ==== 抗劈叉 ====
    double anti=KP_THETA*(leg_R.theta-leg_L.theta)+KD_THETA*(leg_R.theta_dot-leg_L.theta_dot);

    // ==== pass 2: VMC+output ====
    for(int leg=0;leg<2;leg++){
        int isR=(leg==0);
        double phi1=S[leg][0],phi4=S[leg][1],Lv=S[leg][4];
        double phi2=S[leg][7],phi3=S[leg][8],phi5=S[leg][9];
        double T_w_raw=S[leg][10],T_Leg=S[leg][11],F_Leg=S[leg][12];

        T_Leg+=(isR?anti:-anti);
        double rs=isR?1.0:-1.0, fs=isR?-1.0:1.0;
        double J[2][2];
        vmc_jacobian(phi1,phi2,phi3,phi4,phi5,Lv,J);
        double Tj0=J[0][0]*F_Leg+J[0][1]*T_Leg;
        double Tj1=J[1][0]*F_Leg+J[1][1]*T_Leg;

        d_data->ctrl[act_ids[act_idx_map[leg*2]]]  =clamp(rs*(-Tj0),-TORQUE_LIMIT,TORQUE_LIMIT);
        d_data->ctrl[act_ids[act_idx_map[leg*2+1]]]=clamp(fs*(-Tj1),-TORQUE_LIMIT,TORQUE_LIMIT);
        d_data->ctrl[act_ids[2+leg*3]]              =clamp(wheel_pol[leg]*T_w_raw,-TORQUE_LIMIT,TORQUE_LIMIT);
    }
}

// ======================== main ========================
int main(void) {
    if(init_model(XML_PATH)!=0) { fprintf(stderr,"FATAL: init failed\n"); return 1; }

    double phi1_ref,phi4_ref;
    ik(LEG_LENGTH_REF,&phi1_ref,&phi4_ref);
    printf("=== C Controller === L=%.3fm phi1=%.0f phi4=%.0f\n",
           LEG_LENGTH_REF,deg(phi1_ref),deg(phi4_ref));

    double qhr,qfr,qhl,qfl;
    phi_to_q(1,phi1_ref,phi4_ref,&qhr,&qfr);
    phi_to_q(0,phi1_ref,phi4_ref,&qhl,&qfl);
    d->qpos[jnt_qr[0]]=qhr; d->qpos[jnt_qf[0]]=qfr;
    d->qpos[jnt_qr[1]]=qhl; d->qpos[jnt_qf[1]]=qfl;
    mj_forward(m,d);
    build_k_table(0.08,0.45,0.002);

    mjcb_control = my_controller;

    printf("Simulating...\n");
    double last_print=0;
    for(int step=0;step<100000;step++){
        mj_step(m,d);
        if(d->time-last_print>=1.0){
            double pitch=(freejoint_adr>=0)?get_pitch(d,freejoint_adr):0;
            printf("[t=%.1fs] pitch=%.1fdeg  FN_R=%.0fN FN_L=%.0fN\n",
                   d->time,deg(pitch),leg_R.FN,leg_L.FN);
            last_print=d->time;
        }
    }

    mjcb_control=NULL;
    mj_deleteData(d);
    mj_deleteModel(m);
    printf("Done.\n");
    return 0;
}
